### PR TITLE
Support 8 bits PNG file

### LINKED ISSUE #35601: Support palleted PNG file
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

PrestaShop only support true color png file.

### Expected behavior

Support 8 bits png file

### Steps to reproduce

1. Upload a 8 bits png file (256 colors file)
2. Check the uploaded file: it is in true color and size is doubled

### PrestaShop version(s) where the bug happened

8.1

### PHP version(s) where the bug happened

all

### If your bug is related to a module, specify its name and its version

_No response_

### Your company or customer's name goes here (if applicable).

Creabilis.com

### PR DESCRIPTION
**Pull request by jf-viguier**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | 8bits png file are converted to 32bits. With this PR 8bits file remains in 8bits.
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | yes / no
| UI Tests     | https://github.com/florine2623/testing_pr/actions/runs/9402884811 ✅ 
| How to test?      | See issue
| Fixed issue or discussion?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/35601
| Sponsor company   | Creabilis.com

### CODE REVIEW

File: classes/ImageManager.php
Comment (SharakPL) [inline classes/ImageManager.php]:
```suggestion
     * @return int|bool
```

File: classes/ImageManager.php
Comment (matthieu-rolland) [inline classes/ImageManager.php]: could you try this and see if this returns correctly the png color type ? 

```php
ord(@file_get_contents($fileName, NULL, NULL, 25, 1));
```

**Review by matthieu-rolland — COMMENTED** [review]
Adding one file reading for each call of the `resize` function seems like it will sensibly impact performances...

But I didn't find any other way to do it...  we already call `filesize` and an `exif_read_data` but these functions won't give us this information.

If anyone has a better idea I'd be glad to hear it, otherwise I'll approve the pull request once my small feedback has been addressed 👍

**Comment by matthieu-rolland** [comment]
seems like the call of `imagetruecolortopalette` is breaking behat tests in php 8 and above

File: classes/ImageManager.php
Comment (jf-viguier) [reply classes/ImageManager.php]: Thanks, it's fixed @SharakPL

**Comment by jf-viguier** [comment]
@matthieu-rolland the palette issue is fixed in my last commit, please review.
The color palette has to be set before imagesavealpha and imagecolorallocatealpha calls.
Also, imagealphablending is only for true color palette.

**Comment by jf-viguier** [comment]
@matthieu-rolland thanks for the idea, commited.

**Review by Hlavtox — CHANGES_REQUESTED** [review]
I am a bit sceptical about few things…
- Right now all files are unified and have 32bit colors, now we would have some differences.
- It seems that you apply it to all thumbnails. As in the previous point, thumbnails are generated by prestashop in a format we define, we should not look on what the original file was.
- You do this even for webp and avif, is that correct?
- What about the performance impact, you are adding one file read for each thumbnail generated.

Putting block now do it doesnt get accidentaly merged. 👍

**Comment by jf-viguier** [comment]
@Hlavtox 
- It's only for png files, not for webp and avif,
- The size of a 8bits png file is doubled (or more) when converted to 32 bits. We have to care about performance and green aspect
-  you are adding one file read for each thumbnail generated : yes, exif_read_data also and it's in the core just to manage rotation. I read only one beat.

**Comment by jf-viguier** [comment]
@Hlavtox this 4 bits png file is 6Kb, with PrestaShop process, the file is 28Kb after generated in the same size.
coeur-8bits

**Comment by ChillCode** [comment]
When the manager calls getimagesize() it will give the bits on every png file, so something like this should work

```
$image_info = getimagesize(/var/www/html/img/image.png);

$8bits = isset($image_info['bits']) && (int) $image_info['bits'] === 8 ? true : false;

And colortype:

$color_type = isset($image_info['channels']) ? $image_info['channels'] : 'Unknown';
```
Just remembered and found the code I used on a project to know the bits and channels:

```
            $info = getimagesize($file);

            $this->image_info = array(
                'width'     => $info[0],
                'height'    => $info[1],
                'type'      => $info[2],
                'attr'      => $info[3],
                'bits'      => isset($info['bits']) ? $info['bits'] : 8,
                'mime'      => $info['mime'],
                'channels'  => (isset($info['channels'])) ? $info['channels'] : 3
            );

```
so the following lines could be modified:

https://github.com/PrestaShop/PrestaShop/blob/8d8efe81427eec7cfe299dcdefe91679b9691955/classes/ImageManager.php#L209-L223

**Comment by kpodemski** [comment]
It seems @ChillCode provided some valuable insights on how to handle this better.

**Comment by Hlavtox** [comment]
@jf-viguier Well, I see that your code is inside

`if ($destinationFileType == 'png' || $destinationFileType == 'webp' || $destinationFileType == 'avif') {`

It should be executed only if the source and destination is png.

File: classes/ImageManager.php
Comment (jf-viguier) [reply classes/ImageManager.php]: done

**Comment by jf-viguier** [comment]
@Hlavtox thanks, I've fixed it

File: classes/ImageManager.php
Comment (SharakPL) [inline classes/ImageManager.php]:
```suggestion
        if (in_array($destinationFileType, ['png', 'webp', 'avif']) {
```

Comment (jf-viguier) [reply classes/ImageManager.php]: My PR doesn't care about this line of code, thanks.

Comment (SharakPL) [reply classes/ImageManager.php]: 😢

Comment (jf-viguier) [reply classes/ImageManager.php]: @SharakPL don't cry, I've done it for you

Comment (jf-viguier) [reply classes/ImageManager.php]: @SharakPL you have now to convince @Hlavtox that this pr is relevant.

**Comment by ChillCode** [comment]
> @ChillCode the bits property is available only in php 8.2 cf https://www.php.net/manual/fr/function.getimagesize.php

If that was true my software wouldn't work, neither PS:

https://github.com/trippo/ResponsiveFilemanager/blob/51eddae5190cfc4408ade40575ee63404fead0b9/filemanager/include/utils.php#L853-L910

They added the support for avif but PNG worked always.

File: classes/ImageManager.php
Comment (SharakPL) [reply classes/ImageManager.php]: @jf-viguier everybody here knows it's relevant. It's just about doing it right way.

btw please don't name commits after me 🤣

**Comment by ChillCode** [comment]
The bits per channel is present for PNG files but not the channels, and 3 will not always be correct.

As we can see PS uses the bits key to determine the memory required, I used for the same purpose but completely forgot the channels were not present:

https://github.com/PrestaShop/PrestaShop/blob/ff87d16565e5f97db1e1167ba8f07e4860d56bdb/admin-dev/filemanager/include/utils.php#L307-L339

Here they use other method using 
https://github.com/trippo/ResponsiveFilemanager/blob/51eddae5190cfc4408ade40575ee63404fead0b9/filemanager/include/get_png_imageinfo.php

https://github.com/trippo/ResponsiveFilemanager/blob/51eddae5190cfc4408ade40575ee63404fead0b9/filemanager/include/utils.php#L853-L910

I just tested with a 64bits PNG image and worked as expected, it gives me 4 channels instead of 3.

But just afraid that adding another file operation here could be a mess for big sites, when resizing 500.000 can take 16 hours, we will see how long it takes with this change.

EDITED:

Implementing get_png_imageinfo is a good idea, so if image is PNG use this method and get all info using only one file read.

**Comment by matthieu-rolland** [comment]
> another file operation here could be a mess for big sites, when resizing 500.000 can take 16 hours, we will see how long it takes with this change

If you have 500,000 files to process, it's going to take a long time no matter what; you're not exactly worried about saving 10 hours at this point...    

I'm more worried about performance issues when they impact the customer. 

Edit: a suggestion from @jolelievre, let's check the PHP version, if it's PHP >= 8.2 we get the information already fetched from `getimagesize` like @ChillCode suggests, so there is no additional read on the file, and if not, we do it like in this PR. If a merchant has real performance issues because of this feature (which I doubt...) then he can update his PHP to 8.2.

**Comment by SharakPL** [comment]
> If you have 500,000 files to process, it's going to take a long time no matter what; you're not exactly worried about saving 10 hours at this point...

Will be much worse if PS is to correctly show images for each shop and theme of the multistore https://github.com/PrestaShop/PrestaShop/issues/34665

**Comment by jf-viguier** [comment]
@ChillCode thanks for your help.
getimagesize does not support additional image information from PNG files like channels or bits. I've tested with php 8.2 : it returns 8bits on a 32 bits png file and no channel information.
So I'll stay on my actual implementation.
Feel free to implement get_png_imageinfo to optimize on your own.

**Comment by ChillCode** [comment]
Jus edited, we don't need offtopics...

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by Hlavtox** [comment]
@jf-viguier There are two things I would like to challenge in the PR. I am 50/50 it's a good idea.

- The png thumbnails are generated by our software for our purposes. We ignore what was the original file format when generating the png thumbnail, why there should be an exception when dealing with PNGs? Why the output should change depending on the output?
- Then, maybe someone will want to do something with the thumnails. You export them to some XML feeds, download them to your PC or something. Then, what if there are some weird issues with processing some of the pngs suddenly, and you would not know why (because they have different color palette)?
- If points above are decided, then its just the matter of a most performant implementation, with nice comments. :-)

**Comment by SharakPL** [comment]
> * Why the output should change depending on the output?

Did you mean

> Why the output should change depending on the **input**?

It obviously should! If the source file is 8 bit then what's the point in generating much heavier 32 bit thumbs? It would probably be more performant if the original image was used everywhere instead of any thumb version.

**Comment by ChillCode** [comment]
@jf-viguier @matthieu-rolland 

Just tested on a folder with more than 200.000 images and it took few seconds to finish, so there is no performance issues for adding that call.

Thanks for the work!

**Comment by jf-viguier** [comment]
Thanks @ChillCode for the good feedback.
@Hlavtox you still block this PR, any move ?

**Review by nicosomb — DISMISSED** [review]

**Comment by jf-viguier** [comment]
@Hlavtox can you unblock ?

**Comment by Hlavtox** [comment]
@jf-viguier I think you also need a check if the source file is png. Now you are running a check for all file types.

**Comment by jf-viguier** [comment]
> @jf-viguier I think you also need a check if the source file is png. Now you are running a check for all file types.
Please explain. There is a test before the call :
`if ($destinationFileType == 'png')`

**Comment by Hlavtox** [comment]
@jf-viguier You are now checking only the destination file type, but you should also check the source one.

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `deprecations` should be `yes` or `no`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#deprecations))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

**Comment by florine2623** [comment]
Hello @jf-viguier ,

Do you think you could check the previous comment : https://github.com/PrestaShop/PrestaShop/pull/35603#issuecomment-2090173338

Thanks!

**Comment by jf-viguier** [comment]
@Hlavtox I don't understand your comment, the sourceFile is checked : 
`self::getPNGColorType($sourceFile)`

**Comment by florine2623** [comment]
friendly reminder @Hlavtox ^^

**Comment by Hlavtox** [comment]
@jf-viguier Yes, but you are calling `self::getPNGColorType` on all jpgs, all webps, all gifs and other supported files for no reason.

**Comment by jf-viguier** [comment]
@Hlavtox no

the condition is:
`if ($destinationFileType == 'png' && self::getPNGColorType($sourceFile) == 3)`

$destinationFileType == 'png' is evaluated first.
So, for other format than png self::getPNGColorType will not be called.

Explanation here : 
https://www.php.net/manual/en/language.operators.logical.php
"Example #1 Logical operators illustrated"

**Comment by Hlavtox** [comment]
@jf-viguier So, again. You are checking DESTINATION FILE for PNG by `$destinationFileType == 'png'` and source file for nothing, and you call self::getPNGColorType($sourceFile) == 3) on every source file.

**Comment by jf-viguier** [comment]
@Hlavtox my bad, you're right on this.
By the way, calling getimagesize() or exif_imagetype() on a file is certainly more expansive than reanding a single byte in a file.

**Comment by jf-viguier** [comment]
@Hlavtox my bad, getimagesize get file type before, so I can check if the file is in png before.
Please check my last commit

**Review by florine2623 — COMMENTED** [review]
Hello @jf-viguier ,

I tested your PR with the file you sent in the previous comment.

Size of the file I tested : **5KB**

> heart

Size of the file after uploading it to the BO : **4KB**
> heart

Size of the file after uploading it to the BO _with PR_ : **2KB**
> heart_with_pr

Is this the expected behavior ?

Thanks!

**Comment by jf-viguier** [comment]
@florine2623 yes.
In my case, it's 28KB without the PR.

**Review by boherm — APPROVED** [review]

**Comment by florine2623** [comment]
It is QA ✅ !
Thanks!

**Comment by nicosomb** [comment]
@Hlavtox are your requested changes still necessary?

**Comment by jf-viguier** [comment]
@nicosomb no

**Review by Hlavtox — APPROVED** [review]

**Pull request merged**
